07_StagingAndMultipleQueues/main.cpp

```diff
@@ -73,7 +84,414 @@ class StagingAndMultipleQueuesApp final : public examples::BasicMultiQueueApplic
 			return retval;
 		}
 
+		template<typename... Args>
+		void logFailAndTerminate(const char* msg, Args&&... args)
+		{
+			m_logger->log(msg, ILogger::ELL_ERROR, std::forward<Args>(args)...);
+			std::exit(-1);
+		}
+
+	private:
+		std::array<core::smart_refctd_ptr<IGPUImage>, IMAGE_CNT> images;
+
+		smart_refctd_ptr<IGPUBuffer> histogramBuffer = nullptr;
+		nbl::video::IDeviceMemoryAllocator::SMemoryOffset histogramBufferAllocation = {};
+
+		std::mutex imagesReadyMutex;
+		std::condition_variable imagesReady;
+		bool imagesReadyFlag = false;
+
+		std::mutex histogramBufferReadyMutex;
+		std::condition_variable histogramBufferReady;
+		bool histogramBufferReadyFlag = false;
+
+		void loadImages()
+		{
+			IAssetLoader::SAssetLoadParams lp;
+			lp.logger = m_logger.get();
+
+			core::bitflag<IGPUCommandPool::E_CREATE_FLAGS> commandPoolFlags = static_cast<IGPUCommandPool::E_CREATE_FLAGS>(IGPUCommandPool::ECF_RESET_COMMAND_BUFFER_BIT);
+			smart_refctd_ptr<nbl::video::IGPUCommandPool> commandPool = m_device->createCommandPool(getGraphicsQueue()->getFamilyIndex(), commandPoolFlags);
+
+			core::smart_refctd_ptr<IGPUCommandBuffer> transferCmdBuffer;
+			core::smart_refctd_ptr<IGPUCommandBuffer> computeCmdBuffer;
+			m_device->createCommandBuffers(commandPool.get(), IGPUCommandBuffer::EL_PRIMARY, 1u, &transferCmdBuffer);
+			m_device->createCommandBuffers(commandPool.get(), IGPUCommandBuffer::EL_PRIMARY, 1u, &computeCmdBuffer);
```

Comment:
Create `FramesInFlight` commandbuffers and `notify_one` after incrementing the images ready counter after you submit each one of them

Later when you'll have a timeline semaphore per thread, you'll reset the commandbuffer afterthe counter wraps around to the same FramesInFlight value